smelli/unbinned.py

```diff
@@ -0,0 +1,167 @@
+    Methods:
+    - `log_likelihood`: The likelihood as a function of the parameters
+    Instances can be imported and exported from/to YAML using the `load`
+    and `dump` methods.
+    """
+    _input_schema_dict = {
+        'eft': str,
+        'basis': str,
+        'scale': float,
+        'names': list,
+        'samples': str,
+        'weights': str
+    }
+
+    _output_schema_dict = {
+        'eft': str,
+        'basis': str,
+        'scale': float,
+        'names': list,
+        'samples': str,
+        'weights': str
+    }
+
+    def _make_real_acc(i):
+        return lambda wcs: np.real(wcs[i])
+
+    def _make_imag_acc(i):
+        return lambda wcs: np.imag(wcs[i])
+
+    def _complex_to_real(self, values):
+        return list(chain.from_iterable(
+            [np.real(values[i])] if 'real' in self._basis_wcs[n]
+            else
+            [np.real(values[i]), np.imag(values[i])]
+            for i, n in enumerate(self.names)
+        ))
+
+    def __init__(self,
+                 eft,
+                 basis,
+                 scale,
+                 names,
+                 samples,
+                 base_path,
+                 weights=None):
+        """Initialize the instance.
+        Parameters:
+        - `eft`: a string identifying the WET as defined by wcxf
+        - `basis`: a string identifying the WET basis as defined by wcxf
+        - `scale`: a scale (in units of GeV) at which the WET coefficient samples are valid
+        - `names`: an ordered list of names identiying the Wilson coefficients in the WET basis
+        - `samples`: a numpy array of vectors of Wilson coefficients, with each row representing a sample
+        - `weights`: a numpy array of scalar weights on a logarithmic scale for each sample
+        """
+        self.eft     = eft
+        self.basis   = basis
+        self.scale   = scale
+        self.names   = names
+        self.samples = np.load(os.path.join(base_path, samples))
+        self.weights = np.ones(len(self.samples)) if weights is None else np.load(os.path.join(base_path, weights))
+        self.kde     = KernelDensity(algorithm='auto', kernel='gaussian', rtol=1e-4)
+
+        basis = wcxf.Basis[eft, basis]
+        self._basis_wcs = {}
+        for sector in basis.sectors.values():
+            self._basis_wcs.update(sector)
+
+        self.samples = np.array([
+            self._complex_to_real(row) for row in self.samples
+        ])
+        self.kde.fit(X=self.samples, sample_weight=self.weights)
+
+    def log_likelihood(self, par_dict, w):
+        """Return the normalized log-likelihood for all parameters."""
+        if type(w) is Wilson:
+            smeft_wc = w
+        else:
+            smeft_wc = Wilson.from_wc(w.get_wcxf(sector='all', scale=self.scale, par=par_dict, eft='SMEFT', basis='Warsaw'))
+        wet_wc = smeft_wc.match_run(self.scale, self.eft, self.basis)
+        values = _complex_to_real([wet_wc[name] for name in self.names])
```

Comment:
@dvandyk don't you need a `self.` here in front of `_complex_to_real`? The list comprehension you are using in this line could actually be avoided by modifying the `_complex_to_real` method a bit to allow for direct input of a dictionary. This could e.g. be done like

in `__init__`, after the definition of `self._basis_wcs`, define
```python
self._wc_is_real = {n:'real' in self._basis_wcs[n] for n in self.names}
```
use an updated `_complex_to_real` method that can take a dictionary as argument:
```python
def _complex_to_real(self, values):
    if isinstance(values, dict):
        iterator = self._wc_is_real.items()
    else:
        iterator = enumerate(self._wc_is_real.values())
    return list(chain.from_iterable(
        [np.real(values[i])] if real
        else
        [np.real(values[i]), np.imag(values[i])]
        for i, real in iterator
    ))
```

And then the above line could just look like
```python
values = self._complex_to_real(wet_wc)
```

Reply:
@dvandyk, sorry, of course the dictionary `self._wc_is_real` has to be ordered. So one needs to use an `OrderedDict`:
```python
from collections import OrderedDict
```
and then
```python
self._wc_is_real = OrderedDict((n, 'real' in self._basis_wcs[n]) for n in self.names)
```